### ISSUE TITLE
Proposal to replace mplayer with ffmpeg and mpv for crop detection and preview

### ISSUE THREAD

**Issue by donmelton** [donmelton/video_transcoding#123]
# Proposal to replace `mplayer` with `ffmpeg` and `mpv` for crop detection and preview

I've been thinking about doing this for awhile but it wasn't until I got nudged by @legrostdg in #120 that I finally got around to trying it.

While the `mplayer` tool is not moribund yet, the [MPlayer project](http://mplayerhq.hu/) itself is not very active. And I'm concerned that development of it could eventually cease altogether.

Worse, `mplayer` is buggy, slow and unpleasant to use, especially on macOS where it won't even handle fullscreen mode correctly. And for some platforms, it's a pain to get a version that works at all.

In contrast, [`mpv`](https://mpv.io/), which started as a fork of `mplayer`, is a very active project and the tool itself is fast and a pleasure to use. And it's easy to install `mpv` on macOS via Homebrew:

```
brew install mpv
```

Binaries for Windows and Linux are available here:

<https://mpv.io/installation/>

My `detect-crop` tool should be using `mpv` for previewing. And that's an easy enough change to make since `detect-crop` never calls the previewing tool directly but instead prints the commands to use it on the console.

Currently, `detect-crop` prints out something like this to use `mplayer`:

```
mplayer -really-quiet -nosound -vf rectangle=1920:816:0:132 '/path/to/Movie.mkv'
mplayer -really-quiet -nosound -vf crop=1920:816:0:132 '/path/to/Movie.mkv'

transcode-video --crop 132:132:0:0 '/path/to/Movie.mkv'
```

That would become something like this to use `mpv`:

```
mpv --no-audio --vf lavfi=[drawbox=0:132:1920:816:invert:1] '/path/to/Movie.mkv'
mpv --no-audio --vf crop=1920:816:0:132 '/path/to/Movie.mkv'

transcode-video --crop 132:132:0:0 '/path/to/Movie.mkv'
```

And if we get rid of the dependency on `mplayer` for previewing, we should get rid of it for crop detection as well in both the `detect-crop` and `transcode-video` tools.

It turns out `ffmpeg` can do crop detection just fine. In fact, the crop detection code in `mplayer` was originally derived from libraries in `ffmpeg` anyway. So the behavior and results are similar.

But the results are not always identical. The good news is that the few differences I've seen appear to be improvements when using `ffmpeg`.

The other good news is that `ffmpeg` is almost always faster than `mplayer`. Noticeably faster, too.

Since `ffmpeg` is already required by my `convert-video` tool, you should have it installed already. So leveraging it for this duty isn't really adding a new dependency.

While I'll wait for @samhutchins and @JMoVS to weigh in for what this means for the Windows platform since they're the experts there and @ntodd to comment on this about his Docker container, I see no big downsides from this change and would like to check in my already working code soon. Probably before the end of the week.

(BTW, @samhutchins has already given his thumbs up here: https://github.com/donmelton/video_transcoding/issues/120#issuecomment-275638297 .)

Anyway, let me know what you think. Thanks!

**Comment by ntodd** [comment]
👍 Shouldn't be a problem for the docker image.

**Comment by ericcardenas** [comment]
Less dependencies sounds great to me.

**Comment by RodBrown1988** [comment]
👍 Seems like an easy win moving to this.

**Comment by platonium** [comment]
Following along as best I can, this certainly seems like a beneficial change.

Sent from my iPhone

On Jan 30, 2017, at 1:52 PM, Rod Brown <notifications@github.com<mailto:notifications@github.com>> wrote:


? Seems like an easy win moving to this.

-

**Comment by RobertBoyle365** [comment]
Sounds like an improvement to me.

**Comment by JMoVS** [comment]
I'll try to have a run at installing/compiling mpv in lxss and report back tomorrow. Let's see how it goes :D

**Comment by botsmack** [comment]
Sounds good to me too.

**Comment by JMoVS** [comment]
Have you looked at the dependencies for mpv? It's crazy. Like why does mpv *require* youtube-dl? ... I am testing the install right now (and will later upgrade the windows machine to an insider build and test whether bottles work - which would significantly speed up the process), but you are essentially trading mplayer for mpv, whereas mpv has a hell of a lot of depenedencies in brew.

**Comment by legrostdg** [comment]
On debian at least, `mpv` recommends `youtube-dl`, but it does not depend on it.

**Comment by JMoVS** [comment]
ah ok, just looked at https://github.com/Homebrew/homebrew-core/blob/2c0270042339c67eddd58ad9685ac7738827b802/Formula/mpv.rb

Might have been my mistake. It's not that bad

**Comment by donmelton** [comment]
@JMoVS @legrostdg Since, by default, `mpv` is installed via a pre-built Homebrew bottle on macOS, it doesn't seem "heavier" than `mplayer`. But, yes, `mpv` does have more dependencies. Thankfully most of them are optional.

**Comment by donmelton** [comment]
@JMoVS BTW, I'll hold on for your final word since you have to document it. I'm targeting Thursday, two days from now, for the release. In the meantime, I'm continuing to test. Plus, I have to update my own README document for this.

Does anyone want me to check in the code now so they can see how it works? I can easily do that without releasing it.

**Comment by JMoVS** [comment]
@donmelton Well, Windows only as of today recongized the availability of the insider preview build, so I managed to install it today. It seems like Bottles partially work with the windows beta, but I am currently fighting with it, run into a couple of dubious errors.

BTW: If you run `brew deps -n mpv` , it shows a lot of deps for me compared to mplayer.

It will break it for current users on lxss that are not on an insider preview build (aka windows alpha/beta builds) of windows as gawk doesn't compile currently. On the insider preview build, well, I really need some more time as a lot of dependencies aren't bottled (And I'm up to a count of around 25 to 30 dependencies by now) where the majority needs to be built from source. So it's compiling write now, but I have no word on what the result will be yet

**Comment by donmelton** [comment]
@JMoVS I wish I had enough Windows-fu to assist you on this. :(

Stupid question: In the short term, is it possible to install the downloadable `mpv` binary for Windows into LXSS? I realize this is a hack, but based on your description of the current state of the system, it may be weeks or even longer for current users will be able to build `mpv` on their own.

**Comment by donmelton** [comment]
@JMoVS BTW, since I'm proposing that `mpv` be used only for preview and not actually needed for crop detection itself, that means `mpv` is a "soft" dependency. I could, conceivably, add an option to the `crop-detect` tool to print out the preview commands in the old format for `mplayer` instead. Would that bridge the gap on LXSS until we resolve the build issues?

**Comment by legrostdg** [comment]
That seems great: just remove the `mpv`/`mplayer` dependency and print out the preview commands.

**Comment by donmelton** [comment]
@legrostdg Do you mean print out both sets of commands? That might get a little verbose for some folks (including me :) but that's certainly doable in the short term. But it's easier than adding an option.

**Comment by legrostdg** [comment]
No, just print a default set of command (e.g. with `mpv`), and add a "--with-player [mplayer|mpv|vlc?|...]" option.

**Comment by donmelton** [comment]
@legrostdg I like the option name, but `mpv` and `mplayer` are the only tools that I know of capable of drawing the box around the cropping area. You can't do that with VLC. Do you know of another tool that could?

Then again, I could just make `--with-player mpv|mplayer` have only two choices and the default be `mpv`. That's _really_ easy. :)

**Comment by legrostdg** [comment]
`ffplay` may be added there. It would be cool if `vlc` could do that, too!

**Comment by donmelton** [comment]
@legrostdg I hadn't thought of `ffplay`. I'll have to try that. Thanks!

**Comment by samhutchins** [comment]
@donmelton @JMoVS 

## Windows Sitrep
### Windows binaries
mpv is easy to install, [this site](https://mpv.srsfckn.biz/) has builds that follow the release tags of mpv, so the version will match the version in Homebrew. Installation is exactly the same as MPlayer, you extract some exe files to a folder in $PATH

### Bash on Ubuntu on Windows (lxss, linux subsystem)
As @JMoVS has said, using linuxbrew in this system to install mpv doesn't work (although it might in the future). The mpv website does provide a PPA for Ubuntu, and seeing as the Linux subsystem is Ubuntu we can use those to install mpv. But there's no point, you can't use it to preview the crop. There's no way for an application running inside the Linux subsystem to open a graphical window. This is true for MPlayer too, so while MPlayer can be used for detecting the crop it can't be used to preview it.

------

I say go ahead with this change, it reduces the number of 'hard' dependencies, and mpv is much nicer to use than MPlayer in my experience. I would still heavily recommend that Windows users find native, Windows binaries and use those

**Comment by donmelton** [comment]
@samhutchins Sorry I took so long to comment. I've been AFK for the last several hours.

OK, I'll go with @legrostdg 's idea of adding an option to select the preview helper and make that tool an optional install. That'll give @JMoVS time for him or others to resolve the LXSS issues. But I agree with you, the correct solution may be to just install the downloadable binary on Windows, no matter which environment you're using.

**Comment by JMoVS** [comment]
@donmelton Duh, @samhutchins mentioned a thing I didn't even think about yet. 

You don't need to implement the "different player" thing for lxss, as you currently and foreseeably cannot open a player window from within lxss. So yunare actually eliminating one dependency (mplayer) as previewing with mov is not possible. 

BUT to answer your queation: As of the insider build (and the spring update), one will be able to call windows binaries from within linux, so that might mean that calling a windows mpv binary from within lxss might become a possibility. 

Result: Clear go ahead from my side

**Comment by donmelton** [comment]
@JMoVS Sweet! OK, I'll proceed with the aforementioned plan then. But it may not happen for another day or two as the reason I've AFK most of the day is that I am violently ill. So I might not land it until Friday or even later anyway. I will keep you, @samhutchins and @legrostdg posted. In the meantime, take your vitamin C, folks.

**Comment by JMoVS** [comment]
@donmelton Freshly hand pressed orange juice every day helps ;-) Get better soon! And seriously, this is no Apple breakneck piece software. Take your time.

**Comment by donmelton** [comment]
OK, I have finally recovered enough from this vile virus that I think it may be possible for me to take a crack at this today. My thanks for all your well wishes! That, and not opening any other issues during the last week while I was essentially unconscious. :)

**Comment by donmelton** [comment]
**Update:** Folks, I'm now editing the "README" document to reflect all the changes we've discussed here. That and still being sick is why this is taking so long. My apologies.

I've decided not to support other previewing tools besides `mplayer`. Why is that? Because doing so just bloats both the code and the documentation as well as complicating the testing process. I'm sorry, but it's just not worth it to support `ffplay`.

And eventually, I plan to remove support for `mplayer` once we're all confident that `mpv` will work for everyone.

For now, this is how the new option in `detect-crop` will be described:

```
    --player mpv|mplayer
                    select player for preview commands (default: mpv)
```

But I really want to deprecate that option soon. :)
